File: 0205-isomorphic-strings/0205-isomorphic-strings.java
class Solution {
    public boolean isIsomorphic(String s, String t) {
        Map<Character, Character> map = new HashMap<>();
        Set<Character> mapped = new HashSet<>();
        for(int i = 0; i < s.length(); i++){
            if(!map.containsKey(s.charAt(i))){
                if(mapped.contains(t.charAt(i))){
                    return false;
                }
                map.put(s.charAt(i), t.charAt(i));
                mapped.add(t.charAt(i));
            }
            if(map.get(s.charAt(i)) != t.charAt(i)){
                return false;
            }
        }
        return true;
    }
}

File: 0205-isomorphic-strings/0205-isomorphic-strings.py
class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        s_ch_map = {}
        t_ch_map = {}
        
        for i in range(len(s)):
            if s[i] not in s_ch_map:
                s_ch_map[s[i]] = t[i]
            if t[i] not in t_ch_map:
                t_ch_map[t[i]] = s[i]
            
            if s_ch_map[s[i]] != t[i] or s[i] != t_ch_map[t[i]]:
                return False
        
        return True